packages/common-ts/src/base-service/base-service-v2.ts

```diff
@@ -307,30 +318,67 @@ export abstract class BaseServiceV2<
   public async run(): Promise<void> {
     this.done = false
 
-    // Start the metrics server if not yet running.
-    if (!this.metricsServer) {
-      this.logger.info('starting metrics server')
+    // Start the app server if not yet running.
+    if (!this.server) {
+      this.logger.info('starting app server')
+
+      // Start building the app.
+      const app = express()
+
+      // Body parsing.
+      app.use(bodyParser.json())
+      app.use(bodyParser.urlencoded({ extended: true }))
+
+      // Logging.
+      app.use(
+        morgan((tokens, req, res) => {
+          return [
+            tokens.method(req, res),
+            tokens.url(req, res),
+            tokens.status(req, res),
+            JSON.stringify(req.body),
+            '\n',
+            tokens.res(req, res, 'content-length'),
+            '-',
+            tokens['response-time'](req, res),
+            'ms',
+          ].join(' ')
+        })
+      )
 
-      await new Promise((resolve) => {
-        const app = express()
+      // Metrics.
+      // Will expose a /metrics endpoint by default.
+      app.use(
+        promBundle({
+          promRegistry: this.metricsRegistry,
+          includeMethod: true,
+          includePath: true,
+          includeStatusCode: true,
+        })
+      )
 
-        app.get('/metrics', async (_, res) => {
-          res.status(200).send(await this.metricsRegistry.metrics())
+      // Health status.
+      app.get('/healthz', async (req, res) => {
+        return res.json({
+          ok: this.healthy,
         })
+      })
 
-        this.metricsServer = app.listen(
-          this.metricsServerPort,
-          this.metricsServerHostname,
-          () => {
-            resolve(null)
-          }
-        )
+      // Registery user routes.
+      if (this.routes) {
+        this.routes(app)
+      }
+
+      // Wait for server to come up.
+      await new Promise((resolve) => {
```

Comment:
To be able to use supertest to write unit tests like we do herer https://github.com/ethereum-optimism/gateway/blob/develop/packages/backend/src/index.spec.ts#L20

We need app.listen to not get called by default.   My quick hack was to just wrap it in a if(nodenv !== 'test') but to do this cleanly we need to seperate this function into two methods

1. something called something like `private readonly initServer` returns the server
2. Something called something like `private readonly startServer`
3. Then expose a utility seperate from this server that wraps supertest 

import request from 'supertest'
const testBaseService = (service: BaseServiceV2): request.SuperTest<request.Test> => {
    // as any because we are using private method to do this test util
    // alternatively if you prefer you can make it public and throw an error if NODE_ENV is not test.
    request((service as any).initServer())
}

Reply:
In general it's bad practice to call app.listen by default.   You usually want to initialize the app and start the app in two different steps.